File: examples/cpp/my-app/src/services/ProjectService.cpp
#include "services/ProjectService.h"
#include "utils/Logger.h"

namespace services {

ProjectService::ProjectService() {}

std::shared_ptr<models::Project> ProjectService::createProject(const std::string& name, 
                                                               const std::string& description, 
                                                               int ownerId) {
    utils::Logger::getInstance().info("Creating project: " + name);
    auto project = projectRepo.create(name, description, ownerId);
    notificationService.sendNotification(ownerId, "Project created: " + name);
    return project;
}

void ProjectService::updateProject(int projectId, const std::string& name, const std::string& description) {
    utils::Logger::getInstance().info("Updating project: " + std::to_string(projectId));
    projectRepo.update(projectId, name, description);
}

void ProjectService::addMember(int projectId, int userId) {
    utils::Logger::getInstance().info("Adding member to project " + std::to_string(projectId));
    auto project = projectRepo.findById(projectId);
    if (project) {
        project->addMember(userId);
        projectRepo.save(*project);
        notificationService.sendNotification(userId, "You have been added to project: " + project->getName());
    }
}

void ProjectService::removeMember(int projectId, int userId) {
    auto project = projectRepo.findById(projectId);
    if (project) {
        project->removeMember(userId);
        projectRepo.save(*project);
    }
}

void ProjectService::deleteProject(int projectId) {
    utils::Logger::getInstance().info("Deleting project: " + std::to_string(projectId));
    projectRepo.deleteById(projectId);
}

std::vector<models::Project> ProjectService::getProjectsByOwner(int ownerId) {
    return projectRepo.findByOwner(ownerId);
}

std::shared_ptr<models::Project> ProjectService::getProjectById(int projectId) {
    return projectRepo.findById(projectId);
}

} // namespace services

File: examples/python/my-app/src/services/project_service.py
"""Project service with business logic."""

from typing import List, Optional
from models.project import Project
from database.project_repository import ProjectRepository
from utils.logger import Logger
from utils.validator import Validator

class ProjectService:
    """Service for project-related business logic."""
    
    def __init__(self):
        self._repository = ProjectRepository()
        self._logger = Logger.get_instance()
        self._validator = Validator()
    
    def get_project_by_id(self, project_id: int) -> Optional[Project]:
        """Get project by ID."""
        self._logger.info(f"Getting project: {project_id}")
        return self._repository.find_by_id(project_id)
    
    def get_projects_by_owner(self, owner_id: int) -> List[Project]:
        """Get all projects owned by a user."""
        self._logger.info(f"Getting projects for owner: {owner_id}")
        return self._repository.find_by_owner(owner_id)
    
    def get_active_projects(self) -> List[Project]:
        """Get all active projects."""
        self._logger.info("Getting active projects")
        return self._repository.find_active()
    
    def create_project(self, project: Project) -> Project:
        """Create a new project."""
        self._logger.info(f"Creating project: {project.name}")
        
        # Validate
        if not self._validator.is_not_empty(project.name):
            raise ValueError("Project name cannot be empty")
        
        return self._repository.save(project)
    
    def update_project(self, project: Project) -> Project:
        """Update existing project."""
        self._logger.info(f"Updating project: {project.id}")
        return self._repository.update(project)
    
    def archive_project(self, project_id: int) -> Project:
        """Archive a project."""
        self._logger.info(f"Archiving project: {project_id}")
        
        project = self._repository.find_by_id(project_id)
        if not project:
            raise ValueError(f"Project not found: {project_id}")
        
        project.archive()
        return self._repository.update(project)
